### PR TITLE
worker refactor: adds internal packages for worker processing

### PR DESCRIPTION
**Pull request by sunnyagg**
- adds schedule manager for managing scheduling
- adds subscription manager for subscription management

### CODE REVIEW

**Comment by sunnyagg** [comment]
> nit: Move `scheduler` to internal/ dir as its uses `internal/models`
https://github.com/razorpay/metro/pull/198

File: internal/tasks/subscriptionmanager.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+	wg := sync.WaitGroup{}
+	sm.pushHandlers.Range(func(_, handler interface{}) bool {
+		wg.Add(1)
+		ps := handler.(*stream.PushStream)
+		go func(ps *stream.PushStream, wg *sync.WaitGroup) {
+			defer wg.Done()
+
+			err := ps.Stop()
+			if err != nil {
+				logger.Ctx(ctx).Infow("error stopping stream handler", "error", err)
```
Comment (shivasishdas) [inline internal/tasks/subscriptionmanager.go:251]:
```suggestion
				logger.Ctx(ctx).Errorw("error stopping stream handler", "error", err)
```
check other places as well

File: internal/tasks/manager.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package tasks
+
+import "context"
+
+// IManager defines worker interface
+type IManager interface {
```
Comment (shivasishdas) [inline internal/tasks/manager.go:6]: the name sounds a little weird. just call this `Task` instead? we eventually only define `task.run()` with this.
Same with the file name as well?

File: internal/tasks/schedulemanager_test.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/golang/mock/gomock"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+
+	mocks2 "github.com/razorpay/metro/internal/brokerstore/mocks"
+	"github.com/razorpay/metro/internal/common"
+	"github.com/razorpay/metro/pkg/registry/mocks"
+)
+
+func TestScheduleManager_Start(t *testing.T) {
```
Comment (shivasishdas) [inline internal/tasks/schedulemanager_test.go:17]: add UTs for `subscriptionmanager.go`? Do agree with its complexity. Can take it later if needed.

File: pkg/leaderelection/candidate.go [hunk 4]
```diff
@@ -2,28 +2,34 @@ package leaderelection
...
 
 // New creates a instance of LeaderElector from a LeaderElection Config
-func New(model *node.Model, c Config, registry registry.IRegistry) (*Candidate, error) {
+func New(id string, sessionID string, c Config, registry registry.IRegistry) (*Candidate, error) {
 	if err := c.Validate(); err != nil {
 		return nil, err
 	}
 
+	if id == "" {
+		return nil, errors.New("candidate id is nil")
```
Comment (shivasishdas) [inline pkg/leaderelection/candidate.go:20]:
```suggestion
		return nil, errors.New("candidate id is empty")
```

File: pkg/leaderelection/candidate.go [hunk 5]
```diff
@@ -2,28 +2,34 @@ package leaderelection
...
+	}
+
+	if sessionID == "" {
+		return nil, errors.New("sessionID is nil")
```
Comment (shivasishdas) [inline pkg/leaderelection/candidate.go:24]:
```suggestion
		return nil, errors.New("sessionID is empty")
```

File: pkg/leaderelection/candidate_test.go [hunk 6]
```diff
@@ -52,28 +47,23 @@ func TestLeaderElectionRun(t *testing.T) {
...
 	registryMock.EXPECT().Watch(gomock.Any(), gomock.Any()).Return(watcherMock, nil).AnyTimes()
 
 	config := getConfig()
 
 	config.Callbacks.OnStartedLeading = func(ctx context.Context) error {
 		return fmt.Errorf("terminate leader election")
 	}
-	model := getDummyNodeModel()
-	c, _ := New(model, config, registryMock)
+	c, _ := New("node01", "id", config, registryMock)
```
Comment (shivasishdas) [inline pkg/leaderelection/candidate_test.go:59]: Instead of ignoring we can `assert.Nil` on `err` here

File: internal/stream/stream.go [hunk 7]
```diff
@@ -1,4 +1,4 @@
-package worker
+package stream
```
Comment (shivasishdas) [inline internal/stream/stream.go:1]: we have two stream impls. One in `web/` and one in `worker/`. Should this be sub-dir'd accordingly? In future, we can move the web part as well here.

File: pkg/leaderelection/candidate_test.go [hunk 8]
```diff
@@ -96,35 +86,26 @@ func TestDeregisterNodeUnregistered(t *testing.T) {
...
 	assert.NotNil(t, c)
 
-	// there shoudn't be any call to registry since node was not registered
-	c.release(context.Background())
-	assert.Equal(t, "", c.sessionID)
-	assert.Equal(t, false, c.IsLeader())
+	registryMock.EXPECT().Release(gomock.Any(), "id", common.GetBasePrefix()+"leader/election/test", gomock.Any()).Return(true).AnyTimes()
```
Comment (shivasishdas) [inline pkg/leaderelection/candidate_test.go:92]: no need to use `AnyTimes()` here. This is a deterministic call count.

File: pkg/leaderelection/config.go [hunk 9]
```diff
@@ -3,34 +3,20 @@ package leaderelection
...
-
 	// Callbacks are callbacks that are triggered during certain lifecycle
 	// events of the LeaderElector
 	Callbacks LeaderCallbacks
-
-	// Name of the leaderelector service
-	Name string
 }
 
 // Errors raised by package
```
Comment (shivasishdas) [inline pkg/leaderelection/config.go:18]: remove

File: pkg/registry/consul.go [hunk 10]
```diff
@@ -169,6 +169,7 @@ func (c *ConsulClient) Release(ctx context.Context, sessionID string, key string
 	}, nil)
 
 	if err != nil {
+		logger.Ctx(ctx).Errorw("error in consul realease", "ID", sessionID, "key", key, "error", err.Error())
```
Comment (shivasishdas) [inline pkg/registry/consul.go:172]:
```suggestion
		logger.Ctx(ctx).Errorw("error in consul release", "ID", sessionID, "key", key, "error", err.Error())
```
nit typo

File: service/openapi-server/service.go [hunk 11]
```diff
@@ -66,9 +59,29 @@ func (svc *Service) runOpenAPIHandler(_ context.Context) error {
...
+			logger.Ctx(ctx).Infow("http server shutdown failed with err", "error", err.Error())
+		}
+
+	}()
 
 	// Run the server
 	err = server.ListenAndServe()
+
+	if err != nil {
+		logger.Ctx(ctx).Infow("openapi-server returned with error", "error", err.Error())
```
Comment (shivasishdas) [inline service/openapi-server/service.go:83]:
```suggestion
		logger.Ctx(ctx).Errorw("openapi-server returned with error", "error", err.Error())
```

File: service/web/service.go [hunk 12]
```diff
@@ -82,94 +75,70 @@ func (svc *Service) Start(ctx context.Context) error {
...
-		return svc.internalHTTPServer.Shutdown(gctx)
+		err := server.RunInternalHTTPServer(gctx, svc.webConfig.Interfaces.API.InternalHTTPServerAddress)
+		return err
 	})
 
-	err := grp.Wait()
+	err = grp.Wait()
 	if err != nil {
-		logger.Ctx(ctx).Warnw("failed to stop service", "error", err.Error())
+		logger.Ctx(gctx).Info("web service error: %s", err.Error())
```
Comment (shivasishdas) [inline service/web/service.go:139]:
```suggestion
		logger.Ctx(gctx).Errorf("web service error: %s", err.Error())
```

**Review by shivasishdas — CHANGES_REQUESTED** [review]
had a quick look. Will do another pass tomo.

File: internal/stream/stream.go (hunk 7 above)
Comment (sunnyagg) [reply internal/stream/stream.go:1]: Yes, Didn't touch the web part yet but we can move that here slowly

File: internal/tasks/manager.go (hunk 2 above)
Comment (sunnyagg) [reply internal/tasks/manager.go:6]: Yes, Needed suggestions here

File: internal/tasks/schedulemanager_test.go (hunk 3 above)
Comment (sunnyagg) [reply internal/tasks/schedulemanager_test.go:17]: Writing in a separate branch :)

File: pkg/leaderelection/candidate_test.go (hunk 6 above)
Comment (sunnyagg) [reply pkg/leaderelection/candidate_test.go:59]: 👍

File: internal/tasks/manager.go (hunk 2 above)
Comment (shivasishdas) [reply internal/tasks/manager.go:6]: Can we use Task then? As discussed in the future, multiple `Tasks` can be tied together onto one `Workflow` and run by a `Manager`.

Comment (sunnyagg) [reply internal/tasks/manager.go:6]: Yes, Sounds good, but want to keep that in separate PR

Comment (shivasishdas) [reply internal/tasks/manager.go:6]: Sure. Rename at least in this PR? `IManager` really sounds out of place.

Comment (sunnyagg) [reply internal/tasks/manager.go]: Updated

**Review by shivasishdas — APPROVED** [review]

**Pull request merged**